[1 - 2단계 방탈출 예약 대기] 도기(이동엽) 미션 제출합니다.

dye0p opened the pull request:
안녕하세요 샐리! 이번 미션의 리뷰이 도기 입니다.
리뷰어로 만나게 되어서 영광입니다!
부족한 부분이 많은 코드지만 리뷰 잘 부탁 드립니다🙇‍♂️

[커밋 범위](https://github.com/woowacourse/spring-roomescape-waiting/pull/234/files/f042e0949517929f1fe5962d7f79fbd0694ae3df..86b3b19ecd5c392e9ac2be162a6a53d4bf5bf036) 입니다.

<!-- 

## 코드 리뷰 팁

- 코드와 관련된 질문이 있다면, PR 본문에 적기 보다는 해당 코드를 선택하고 코멘트를 남겨주세요.
  - [참고: Adding comments to a pull request](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/commenting-on-a-pull-request#adding-comments-to-a-pull-request)

-->
## 체크 리스트
- [x] 미션의 필수 요구사항을 모두 구현했나요?
- [x] Gradle `test`를 실행했을 때, 모든 테스트가 정상적으로 통과했나요?
- [x] 애플리케이션이 정상적으로 실행되나요?
- [x] [프롤로그](https://prolog.techcourse.co.kr)에 셀프 체크를 작성했나요?
  <!-- 작성한 셀프 체크의 링크를 남겨주세요. -->
   - [셀프 체크](https://prolog.techcourse.co.kr/studylogs/4373)

## 객체지향 생활체조 요구사항을 얼마나 잘 충족했다고 생각하시나요?
### 1~5점 중에서 선택해주세요.

- [ ] 1 (전혀 충족하지 못함)
- [x] 2
- [ ] 3 (보통)
- [ ] 4
- [ ] 5 (완벽하게 충족)

### 선택한 점수의 이유를 적어주세요.
<!-- 이유 작성 -->

JPA와 스프링을 사용하면서 객체지향적으로 구현하지 못한 것 같습니다. 클래스 인스턴스가 3개가 넘어가는 경우가 많았고, getter 사용을 무분별하게 하였습니다.

## 베이스 코드 선택 체크
- [ ] 이전 미션의 내 코드에서 시작 
- [x] 이전 미션의 페어의 코드에서 시작

## 어떤 부분에 집중하여 리뷰해야 할까요?
<!-- 리뷰어가 효과적으로 피드백할 수 있도록 중점적으로 리뷰 받고 싶은 내용을 *요약* 형태로 작성해 주세요.
리뷰해야 할 포인트를 요약해 강조해 주시면, 리뷰어가 코드 전체를 이 부분에 집중해 리뷰할 수 있습니다.
반면 특정 코드부분에 대한 피드백이 필요하다면, 이 곳에 적기 보다 해당 코드를 선택하고 코멘트를 남기는 것을 권장합니다. -->

### JPA와 객체지향
ORM은 분명 데이터 베이스와 애플리케이션 코드를 객체지향적으로 사용할 수 있도록 도와주는 개념이지만, 기존의 코드를 JPA로 마이그레이션 하면서 객체지향적으로 코드가 변했다는 느낌을 느끼기 힘들었습니다. 아직 JPA를 사용하는 것이 미숙하여 판단하기가 어렵습니다. 제가 JPA를 올바르게 사용하고 있는지, 주의해야 할 부분이나, 학습하면 좋을 키워드나 개념이 있다면 피드백 주시면 좋을 것 같습니다!

### 연관관계 판단의 어려움
엔티티를 설계하면서 테이블 간의 관계를 객체로 자연스럽게 표현하는 것이 쉽지 않았습니다. 특히 양방향 매핑을 언제 사용하는 것이 적절한지, 그리고 테이블 간의 관계를 JPA에서 엔티티 간의 관계로 어떻게 옮겨와야 하는지에 대한 판단 기준이 명확하지 않았습니다. 이런 기준에 대한 감을 익히기 위해 학습하면 좋을 키워드나 개념이 있다면 피드백 주시면 좋을 것 같습니다!


### 📌 코드리뷰 질문 그라운드룰 (지우지 마세요)
✅ 많은 사람과 공유하고 토론해보는 것이 더 좋은 질문에는, 리뷰어가 💬말풍선 이모지를 남겨 토론을 추천할 수 있어요.  
✅ 토론을 추천받은 질문은 `#be-7기-잡담-질답`슬랙채널이나 캠퍼스에서 자유롭게 토론을 이어가보세요!  
✅ 코드에 대한 맥락 없이 리뷰어의 주관적인 의견을 묻거나, 단순히 현업에서는 어떻게 하는지를 묻는 질문은 코드 리뷰 본문에는 어울리지 않아요.  
코드 리뷰는 변경된 코드의 목적, 의도, 품질을 중심으로 논의하는 자리예요.  
✅ 코드와 관련된 질문이 있다면 PR 본문에 적기보다는, 해당 코드를 선택해 코멘트를 남겨주세요.

Jin409 committed:
chore: 0단계 준비 완료

Jin409 committed:
chore: 0단계 준비 완료

dye0p committed:
Merge remote-tracking branch 'origin/step1' into step1

# Conflicts:
#	src/main/java/roomescape/dao/ReservationDao.java
#	src/main/java/roomescape/dao/ReservationJdbcDao.java
#	src/main/java/roomescape/presentation/controller/AuthController.java
#	src/main/java/roomescape/presentation/controller/ReservationAdminController.java
#	src/main/java/roomescape/presentation/controller/ReservationController.java
#	src/main/java/roomescape/presentation/controller/ReservationTimeController.java
#	src/main/java/roomescape/presentation/controller/ThemeController.java
#	src/main/java/roomescape/service/ReservationService.java
#	src/test/java/roomescape/dao/ReservationJdbcDaoTest.java
#	src/test/java/roomescape/fake/FakeReservationDao.java

dye0p committed:
docs: 1단계 요구사항 업데이트

dye0p committed:
chore: JPA 의존성 추가

dye0p committed:
chore: lombok 의존성 추가

dye0p committed:
feat: 엔티티 매핑

dye0p committed:
refactor: JPA를 활용하여 데이터베이스에 접근하도록 수정

dye0p committed:
docs: 2단계 요구사항 업데이트

dye0p committed:
feat:  내 예약 목록을 조회하는 API를 구현한다

dye0p committed:
test: 사용자의 예약 내역 조회 테스트 추가

dye0p committed:
test: 사용자의 예약 내역 조회 인수 테스트 추가

dye0p committed:
chore: 사용하지 않는 의존성 제거

dye0p committed:
refactor: 코드 메서드 분리 밎 정렬

dye0p committed:
test: ThemeService 테스트 추가

dye0p committed:
remove: 사용하지 않는 dao 제거

dye0p committed:
fix: findTopReservedThemeSince 쿼리 수정

dye0p committed:
remove: AvailableReservationTime 객체 제거

dye0p committed:
test: ThemeRepository 인기테마 조회 테스트 추가

dye0p committed:
remove: 사용하지 않는 dto 제거

dye0p committed:
refactor: 요청 검증 Spring validation을 이용하도록 변경

dye0p commented on src/main/java/roomescape/service/JwtProvider.java:
```diff
@@ -0,0 +1,76 @@
+import org.springframework.stereotype.Service;
+import roomescape.common.exception.UnauthorizedException;
+
+@Service
+public class JwtProvider {
```
**[@Service 은 적절한가?]**
`JwtProvider` 는 `AuthService`에서 사용하고 있습니다. @Service 는 비즈니스 로직을 수행하는 성격이 강한 객체에게 사용하는 것 이라고 생각합니다. 하지만 JwtProvider는 비즈니스 로직의 성경을 띄기보단 유틸성 객체에 더 가깝다는 생각이듭니다. 따라서 @Component를 사용하는 것도 적절한 선택이라고 생각이드는데, 이에 대한 샐리의 의견이 궁금합니다!

dusdn1702 replied:
요 코멘트를 놓쳤네요.
저라면 Component를 사용할 것 같아요.
말씀 주신 것처럼 도메인을 기반으로 비즈니스 로직을 수행하는 것이 아닌 유틸성 객체이기 때문이에요.
`@Service` 어노테이션을 언제 사용할지에 대해서 생각해보고, 주위 크루, 문서도 찾아보며 도기만의 기준을 만들어보고 공유해주세요!

dye0p replied:
저는 `@Service` 어노테이션에 대해서 빈으로 등록하는 것 이상의 의미를 담고 있다고 생각합니다. 어노테이션을 통해서 해당 객체는 애플리케이션의 비즈니스 로직을 수행하는 객체라는 인식을 줄 수 있다고 생각하기 때문입니다! 

따라서, 지금처럼 비즈니스 로직을 수행하거나, 트랜잭션 범위의 로직을 수행하는 객체라면 `@Service` 어노테이션을 사용할 것 같습니다!

dusdn1702 replied:
좋습니다~~ 👍

dusdn1702 commented on src/main/java/roomescape/model/Member.java:
```diff
@@ -0,0 +1,43 @@
+public class Member {
+
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
```
Id를 이렇게 생성되도록 한 이유는 무엇인가요? 다른 방식에는 어떤게 있었을까요?

dye0p replied:
JPA의 **기본키 생성 전략**은 크게 두가지가 있습니다. 
- 직접 할당: 개발자가 직접 애플리케이션에서 기본 키를 할당한다.
- 자동 생성: 개발자가 직접 할당하지 않는다 (DB에게 위임하거나, 키 생성 전용 테이블을 생성할 수 있다)

이번 미션에선 `자동 생성` 방식을 사용했습니다. `직접 할당` 방식을 사용하지 않는 이유는 다음과 같습니다.

- 개발자의 실수로 할당을 누락할 수 있습니다.(식별자 지정을 누락하고 `persist`를 하게되면 예외가 발생하긴 합니다.) 이로인해 개발자가 신경써야 할 것이 늘어납니다.

- db의 데이터는 식별자(ID)로 식별합니다. 개발자가 직접 할당하는 경우 잘못된 식별자를 할당하면 이미 존재하는 식별자와 충돌할 수 있습니다. ex) `member.setId(100L)`

- 실제 프로젝트 환경에서 대부분의 기본키 생성 전략은  `자동 생성` 방식을 사용합니다. 지금은 아니지만, 협업을 할 때 다같이 직접적인 키 생성 방식에 신경을 쓰기보단, 자동 생성 방식에 맡기는 것이 효율적이라 생각합니다.

이러한 이유로 직접 할당을 하지 않고, `자동 생성` 방식을 선택했습니다!

자동 생성 방식에도 여러가지 방식이 있지만 `IDENTITY` 방식을 사용한 이유는 다음과 같습니다.
- `TABLE` 생성 방식은 불필요한 테이블을 생성한다 생각했고,  `IDENTITY` 보다 나은 점이 없다고 판단했습니다.
- `SEQUENCE` 방식은 객체가 영속화 되기 전, (persist 전) 식별자(ID)가 부여됩니다.
`IDENTITY` 방식은 객체가 영속화 된 후 (persist된 후) 식별자(ID)가 부여됩니다. 개인적인 생각이지만, 영속화 되기 전 ID가 필요한 경우는 드물다고 생각합니다! 따라서 SEQUENCE 방식보다 IDENTITY 방식을 사용하였습니다.

키 생성 방식을 공부하다 `AUTO` 전략이 있다는 것을 알게되었습니다. 이는 사용하는 데이터베이스에 따라서 생성 전략중 하나를  자동으로 선택해줍니다. 하지만 `IDENTITY` 방식처럼 특정 생성 전략을 명시해주면 테이블을 보고도 어떤 db를 사용하는지 알 수 있다고 생각합니다!

**식별자가 IDENTITY로 할당된 객체의 식별자를 setId()로 변경하는 경우**
기술적으로 가능하지만 식별자는 불변해야 하기에 절대 해서는 안된다고 생각합니다.  따라서 ID를 변경할 수 있도록 setter()를 열어주는 것을 주의해야할 것 같습니다.

dusdn1702 replied:
좋습니다 👍

dusdn1702 commented on src/main/java/roomescape/model/Member.java:
```diff
@@ -0,0 +1,43 @@
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+
+    private String name;
```
column 선언을 별도로 안한 이유는 무엇인가요?
추가하면 어떤 변화가 있을까요?

dye0p replied:
기존의 테이블에서 각각의 컬럼에 `NotNull` 제약조건과 `Member` 테이블의 `email` 컬럼같은 경우 `unique` 제약조건이 있었습니다. 이러한 제약조건 설정을 누락한 것 같습니다.

`Colunm`과  `JoinColumn`을 통해서 각 컬럼의 `NotNull` 제약 조건과 `Unique Key` 제약조건, 컬럼 Name 등등 다양한 설정을 할 수 있습니다!

전
``` sql
Hibernate: 
    create table member (
        id bigint generated by default as identity,
        email varchar(255),
        name varchar(255),
        password varchar(255),
        role enum ('ADMIN','USER'),
        primary key (id)
    )
```

후
``` sql
Hibernate: 
    create table member (
        id bigint generated by default as identity,
        email varchar(255) not null unique,
        name varchar(255) not null,
        password varchar(255) not null,
        role enum ('ADMIN','USER') not null,
        primary key (id)
    )
```
[refactor: 엔티티 컬럼 매핑 제약조건 설정](https://github.com/woowacourse/spring-roomescape-waiting/pull/234/commits/4c25d5bedf892b15380abbc18a81eaec93059696)

dusdn1702 commented on src/main/java/roomescape/model/Member.java:
```diff
@@ -0,0 +1,43 @@
+@Entity
+@Getter
+@AllArgsConstructor
+@NoArgsConstructor(access = AccessLevel.PROTECTED)
```
이 어노테이션은 왜 필요했나요?

dye0p replied:
#### Entity
객체를 엔티티로 인식 시키기 위함이었습니다.

#### Getter
DTO <-> Entity 변환시 DTO에서 엔티티를 주고 값을 꺼내서 사용하기 위해 getter가 필요하였습니다. 코드로 getter를 작성해주게 되면 코드가 길어지기 때문에 getter 어노테이션을 사용하게 되었습니다.

#### NoArgsConstructor
JPA에서 엔티티를 생성할때 기본적으로 리플렉션을 활용하여 생성합니다. 이때 리플렉션을 위해서 기본 생성자가 필요합니다. 자바에서 별도의 생성자가 없다면 묵시적으로 기본 생성자를 할당하지만, 별도의 생성자가 있다면 기본 생성자를 직접 정의해주어야 합니다.
NoArgsConstructor 를 사용하면 매개변수가 없는 기본 생성자를 만들어 주기 때문에 코드를 작성하지 않아도 된다는 장점이 있기 때문에 사용하게 되었습니다. 
AccessLevel을 PROTECTED로 설정해준 이유는 JPA가 리플렉션을 사용할 수 있게 하고, 외부에서 new 생성을 제한할 수 있다는 장점과 JPA의 권장 방식이기 때문이었습니다. 

#### AllArgsConstructor
테스트 코드 작성시 모든 필드를 매개변수로 받는 객체를 생성하기 위해서 사용했습니다. 테스트를 성공시키기 위해서 별 다른 생각없이 **테스트를 위한 프로덕션 코드**를 추가했네요.. `AllArgsConstructor`에 대해서 추가로 알아보니, 객체 생성시 생성자의 매개변수 순서에 영향을 받는 다는 단점이 있다는 것을 알게 되었습니다. 이는 롬복을 사용하기 전, 자바의 생성자 방식에 대한 불편함이기도 했습니다. 

하지만 가장 중요한 테스트 코드를 위한 프로덕션 코드의 추가가 좋지 않다는 점을 생각해서 제거하였습니다!

[refactor: 사용하지 않는 AllArgsConstructor 어노테이션 제거](https://github.com/woowacourse/spring-roomescape-waiting/pull/234/commits/2c12aed05a4262a7700a810594209a912325a611)

dusdn1702 replied:
좋습니다. 
`@NoArgsConstructor` 도 사용 시 어떤 주의할 점이 있을까요?
`@Getter`도 필요한 필드만 함수로 만들어주는 것은 어때요? 어노테이션을 택한 이유가 무엇인가요?

dye0p replied:
> @NoArgsConstructor 도 사용 시 어떤 주의할 점이 있을까요?

우선 자바에서 별도의 생성자를 지정해주지 않으면 보이진 않지만, 자동으로 기본 생성자를 만들어줍니다.
하지만 기본 생성자를 통해서 발생할 수 있는 문제가 있습니다.

가장 먼저 생각해볼 수 있는 문제는 기본 생성자가 `public`으로 열려있기 때문에 외부에서 `new Member()`를 호출할 수 있습니다.
``` java
    @Transactional
    public void saveReservation(ReservationAdminRegisterDto registerDto) {
        Member member = findMemberById(registerDto.memberId());
        member = new Member(); // 필드가 null인 member 객체로 재할당된다.
        ...
}
```
이러한 상황이 발생할 수 있습니다. 이를 해결하기 위해선
``` java
fianl Member member = findMemberById(registerDto.memberId());
```
이렇게 반환 변수에 `final` 키워드를 통해 재할당을 막을 수도 있습니다.
하지만, 개발자가 `final` 키워드를 항상 붙여줘야하는 불편함과, 누락할 수 있습니다. 그래서 근본적인 해결방법이 되지 않습니다.

그래서`@NoArgsConstructor(access = AccessLevel.PROTECTED)` 옵션을 통해서 생성자의 접근권한을 `protected`로 설정하여 같은 패키지를 제외한 외부에서 생성자를 호출하지 못하도록 제한 할 수 있습니다. 
`protected` 특성상 같은 패키지의 외부에서는 여전히 호출 할 수 있습니다. 이 부분은 어쩔 수 없는 트레이드 오프라고 생각합니다..

**🤔 그럼 차라리 `private` 로 막으면 안되나요?**
이건 JPA가 리플렉션을 이용하여 엔티티를 생성하기 때문에 기본 생성자의 접근제어자가 `public`또는 `protected`여야 하는 제약이 있습니다. 

그래서 엔티티를 외부에서 최대한 보호하기 위해서 기본 생성자의 접근제어자를 `protected`로 설정해야 한다고 생각합니다.
하지만, 엔티티에 별도의 매개변수를 받는 사용자 지정 생성자가 있는 경우, 외부에서 이 생성자를 호출 할 수 있기 때문에 

```java
    @Transactional
    public void saveReservation(ReservationAdminRegisterDto registerDto) {
        Member member = findMemberById(registerDto.memberId());
        member = new Member("도기", "email@", "1234", Role.ADMIN);
        ...
}
```
지역변수로 할당되는 변수는 `final` 키워드를 붙여주는 것이 좋아보입니다. 

### 결론
JPA는 사용자 정의 생성자와 별도로 기본 생성자가 필수입니다. 
그래서 `@NoArgsConstructor`는 기본 생성자를 직접 명시하지 않아도, 만들어주기 때문에 간편하게 사용할 수 있습니다.
하지만 생성자의 접근제어자가 열려있다면 문제가 발생할 수 있기 때문에 접근레벨을 `protected`로 설정해야 하고, 이외에 생성자가 필요하다면 `정적 팩터리 메서드`를 도입하는 것이 좋다고 생각합니다!

dye0p replied:
> @Getter도 필요한 필드만 함수로 만들어주는 것은 어때요? 어노테이션을 택한 이유가 무엇인가요?

처음엔 `DTO <-> Entity` 변환 과정에서 엔티티의 `getter`가 필요할 것 같아, 우선 클래스 단위에 `@Getter`를 사용했습니다. 메서드를 작성하지 않고, 어노테이션을 사용한 이유는 지저분한 `getter` 메서드를 보고싶지 않았습니다. 하지만 이건 필요에 의해서 사용했다기 보단, 그럴 것이다 라고 예측하여 사용했다는 느낌이 강한 것 같습니다. 

`@Getter`를 클래스 레벨에 사용한 경우에 정말 필요하지 않은 필드에 대해서도 `getter`가 열려있다는 것은 다른 개발자가 봤을 때 "사용해도 좋겠다." 라는 인식을 심어줄 것 같습니다. 그래서 필요한 필드에만 `getter`를 사용하도록 수정했습니다! 

[refactor: getter가 필요한 필드에만 getter 적용](https://github.com/dye0p/spring-roomescape-waiting/commit/b50333d68233533c09fe8f9496315609037a7ec2)

`getter` 메서드를 작성해도 상관없지만, 필드에 `@Getter`를 명시해주었습니다. 이렇게 했을 때의 장점은 어노테이션으로 인해서 필드만 봤을 때 `getter` 사용이 가능하다는 것을 쉽게 확인 할 수 있을 것 같습니다. 하지만 리스트를 필드로 가지고 있을 땐 직접 `getter` 메서드를 정의하여 `불변으로 반환`하는 등의 조치를 취해야 할 것 같습니다!

### 결론
- getter 사용은 필요한 필드에만 사용하는 것이 좋을 것 같다.
- @Getter를 필드에 사용하는 것은 개인적인 기준에선 나쁘지 않은 것 같다.
- 필드가 리스트인 경우는 직접 getter 메서드를 작성하여 불변으로 반환하자.

**getter 사용에 대한 개인적인 생각**은
`지금의 getter 사용 방식은 적절한가?`입니다. 레벨 1에서 배운 `getter는 지양하는 것이 좋다`는 무조건 지양하는 것이 아니라, `도메인의 비즈니스적 흐름을 제어하기 위해서 꺼내 쓰지말라`로 이해 했습니다. 그래서 조금 과하게 `getter`를 사용하지 않으려했고, 실제로 쓰일 일이 많지 않았습니다. 하지만 지금은 정말로 외부에서 필요하기 때문에 사용하고 있습니다. 이런 관점에 대해서 샐리는 어떻게 생각하시나요?

추가로, 샐리는 레벨1에서 레벨2로 넘어가면서 배운것들에 대한 괴리감을 어떻게 이해하고 해소하셨는지 궁금합니다!

dusdn1702 replied:
맞아요. ㅎㅎ
점점 레벨이 지나고 프로젝트가 복잡해질수록 이 괴리는 더 깊어질텐데요.
레벨1에서 학습했던 내용은 객체지향을 학습하며 기반을 다지는 과정이었고, 그것을 프로젝트에 적용해가면서 `좋은 코드란 무엇인가` 에 대해 꾸준히 고민하며 나만의 기준을 확립해가는 과정이라고 생각했으면 좋겠어요.
모든 코드에 정답은 없으니까요!

dusdn1702 commented on src/main/java/roomescape/model/Reservation.java:
```diff
@@ -0,0 +1,47 @@
+
+    private LocalDate date;
+
+    @ManyToOne
```
연관관계를 ManyToOne으로 하신 이유가 무엇인가요?
다대일 사용 시 주의할 점은 무엇이 있을까요?

dye0p replied:
"**하나의 예약시간은 여러개의 예약에 포함될 수 있다**" 의 관계로 볼 때 
예약시간 테이블과 예약 테이블의 관계는 `1:N` 관계로 볼 수 있습니다. 따라서 `Reservation`이 'N'이 되고, `ReservationTime`은 1이 됩니다. 데이터베이스에서 외래키는 두 테이블 중 "N"인 테이블이 가지게 됩니다. (N에서 1을 조회할 수 있기 때문) 따라서 `Reservation` 엔티티는 `ReservationTime`의 id를 외래키로 가지게 됩니다! 

위와같은 이유로 `Reservation` 엔티티에서 `ReservationTime`을  `ManyToOne`관계로 설정해주었습니다!

## 주의 사항
### 다대일 사용시 JPA는 ManyToOne을 사용합니다. 이때 외래키의 제약조건을 NotNull로 설정해주는 것이 좋습니다. 

### 즉시로딩 사용에 주의해야합니다.
**ManyToOne은 기본 로딩 전략이 즉시로딩 전략(EAGER) 입니다.** 
즉시로딩은 연관관계의 테이블을 한번에 `join`하여 조회합니다. 만약, 어떠한 데이터를 조회하고 싶을 때 즉시로딩 방식은 연관되어 있는 모든 테이블을 한번에 `join`하여 조회하기 때문에, 필요하지 않은 데이터도 조회되면 불필요한 조회가 발생하게 됩니다. 이는 `N + 1` 문제로도 이어질 수 있다고 하는데, 이 부분은 조금 더 공부를 해보겠습니다!
`ManyToOne`의 기본 전략이 즉시 로딩인 이유는 `ManyToOne`의 관계에서 조회의 방향은 `다 -> 1 `로 조회합니다. 
즉, 많지않은 데이터를 조회하기 때문에(예약 시간을 하나만 조회) 즉시 로딩해도 부담이 적지 않습니다. 또한, 연관된 데이터를 함께 사용하는 경우가 많습니다.

**이를 해결하기 위해서 지연 로딩 전략(LAZY)을 사용할 수 있습니다.** 
지연 로딩 전략은 조회가 필요한 엔티티를 직접 사용하는 시점에서 조회합니다. 즉시 로딩과는 다르게 필요한 시점에서 데이터를 조회하기 때문에 불필요한 조회없이 효율적입니다. 

연관된 데이터를 함께 사용하는 상황이라면 한번에 조회하는 즉시 로딩이 효율적일 수 있다고 생각합니다. 하지만 모든 상황에서 연관 데이터를 함께 사용하지는 않으며, 예상하지 못한 쿼리가 발생할 수 있기 때문에 조회 시점을 개발자가 통제할 수 있는 지연 로딩이 조금 더 나은 선택이라고 생각합니다!

### 양방향 매핑시 연관관계에 주의해야 합니다.
양방향 연관관계시 연관관계의 주인은 `ManyToOne`을 사용하는 쪽입니다. 연관관계의 주인만 외래키를 가지고 반대쪽의 값을 변경할 수 있기 때문에 `OneToMany`를 사용하는 쪽에선 반대쪽의 값을 변경해도 db에 반영되지 않습니다. (반대쪽에서 참조하는 객체의 값을 변경해도 실제 db에 업데이트 쿼리문이 나가지 않는다. 이는 JPA에서 연관관계의 주인만 영속성 컨텍스트에서 관리하기 때문 입니다!)
양방향 매핑시 연관관계의 주인 쪽에서만 값을 세팅할 수 있도록 주의해야 합니다.

[refactor: 연관관계 로딩 전략 번경(즉시 로딩 -> 지연 로딩)](https://github.com/woowacourse/spring-roomescape-waiting/pull/234/commits/9bdc32e4be2398cbd3e3264294d970c99f0aa1ff)

dusdn1702 replied:
전략 적용까지 👍 
양방향 매핑을 적용하지 않은 이유가 궁금해요. 양방향으로 적용했을 때의 장단점에 대해서도요! 
직접 코드로 적용해서 체득한 장단점이면 더더욱 좋을 것 같네요. 😸

dye0p replied:
양방향 매핑을 적용해보고, 제가 느꼈던 양방향 매핑을 적용하지 않은 다음과 같습니다.

### 비즈니스 요구사항에서 **양방향 참조가 필요하지 않기 때문**

실제 DB에서 양방향 매핑은 두개의 단방향 관계입니다. JPA에서는 두개의 엔티티가 서로 참조하는 형태인데, 한쪽에서 외래키를 관리하기 위해서는 연관관계의 주인을 명확히 구별해주어야 합니다. 연관관계의 주인을 명확하게 관리하지 않으면 실수가 발생할 수 있습니다.

특히 연관관계의 주인이 아닌 객체에만 값을 설정할 경우, 실제 DB에는 반영되지 않습니다 왜냐하면 영속성 컨텍스트는 연관관계의 주인만 외래키에 관한 설정을 할 수 있도록 하기 때문입니다.

양방향 매핑을 적용하고 테스트를 했을 때, `ReservationTime`의 `add()` 메서드에서 연관관계 주인인 `Reservation` 쪽의 필드를 설정하지 않아, `reservation` 테이블의 `reservation_time_id` 조회 결과가 `null`로 나오는 문제가 있었습니다.
이를 방지하려면 `add()` 메서드에서 양쪽의 연관관계를 모두 설정해줘야 합니다.

### 양방향 매핑시 **연관관계의 주인이 아닌 쪽에도 변경된 값 설정을 해줘야 한다**
DB상으로는 하나의 테이블에서 외래키를 관리하지만, JPA에서는 두개의 객체가 서로 참조하고 있기 때문에, 한쪽에서 변경이 일어나면 다른 한쪽에도 반영을 해줘야 합니다. 이건 개발자의 입장에서 신경 쓸 것이 더 늘어날 수 있겠다 라는 생각이 들었습니다.

또 다른 단점으론,

### 순환 참조 문제
양쪽이 서로를 참조하고 있으면 JSON 직렬화 과정에서 순환 참조 문제가 발생할 수 있다고 합니다!

이러한 단점을 느끼고, "**우선 단방향 매핑을 하고, 양방향 매핑이 꼭 필요한 상황에서만 하는 것이 좋겠다**" 라는 결론을 내렸습니다! 하지만 양방향 매핑의 장점도 있다고 생각합니다.

- 양쪽에서 조회 가능
    - `예약 -> 예약시간 조회` 가능
    - `예약시간 -> 예약 조회` 가능
- 양쪽에서 데이터를 동기화 할 수 있다
   - 양쪽에서 데이터를 동기화 하기 위해서는 위에서 말한 단점이 있을 수 있다고 생각하는데, 이건 연관관계 편의 메서드를 통해서 어느정도 해소 할 수 있다고 합니다! 

그래서 우선, 지금 구조에서는 단방향 매핑만으로도 충분하다고 판단했고, 불필요한 복잡도를 피하기 위해 단방향 매핑을 선택했습니다.

아직 많은 연관관계를 경험 해보지 못해서 어떤 상황에서 적절한 매핑 관계를 이룰 수 있을지 확실한 판단을 내리기는 어려운 것 같습니다..! 제가 느낀 양방향 매핑의 장단점을 샐리는 어떻게 생각하시나요? 혹시나 보충할 내용이 있다면 피드백 해주시면 감사하겠습니다 : )

dusdn1702 replied:
너무 좋은 것 같습니다 👍 
단순히 기피하는 것이 아니라 필요한 상황이 있음을 느꼈으면 해서 말씀드려봤어요.
이러한 단점을 어떻게 해소할 수 있을지도 같이 고민해보며 적재적소에 적용할 수 있는 도기가 되었으면 좋겠습니다 😄

dusdn1702 commented on src/main/java/roomescape/repository/ReservationRepository.java:
```diff
@@ -0,0 +1,25 @@
+
+    Optional<Reservation> findByDateAndReservationTime(LocalDate date, ReservationTime time);
+
+    List<Reservation> findByTheme_IdAndMember_IdAndDateBetween(
```
_ 언더바는 꼭 필요했을까요??

dye0p replied:
JpaRepository의 쿼리 메서드를 생성하는 과정에서 생겨난 네이밍을 그대로 사용한 것 같습니다! 자바 네이밍 컨벤션에 맞도록 수정했습니다!

[refactor: repository 메서드 네이밍 컨벤션에 맞게 수정](https://github.com/woowacourse/spring-roomescape-waiting/pull/234/commits/a7bff7872e15185f30c008e7b3598c82427b9392)

dusdn1702 commented on src/main/java/roomescape/repository/ThemeRepository.java:
```diff
@@ -0,0 +1,31 @@
+    boolean existsByName(String name);
+
+    @Query(value = """
+            SELECT t.*
```
query를 이용하신 이유가 궁금해요.
다른 방식은 무엇이 있을까요? native query를 이용하면 어떤 단점이 있을까요?

dye0p replied:
JPA에서 직접 sql을 다뤄야 할 때 어떤 방식으로 다루는지 제대로 알지 못했습니다. 그래서 우선 쿼리문을 다루기 위해서 native query를 사용하였습니다! 코드를 다시보니 사용방식을 자세하게 알지 못하고 사용했네요..!

native query를 사용한 이유는 JPQL의 사용방식을 제대로 학습하지 못했습니다. 그래서 빠르게 구현을 하고자 이전 미션에서 사용하던 쿼리문을 그대로 사용하게 되었습니다.

native query 사용시 제가 느꼈던 장단점을 정리해보았습니다.
장점
- 쿼리문을 직접 사용하기 때문에 복잡한 연산도 쿼리로 표현가능하다.

단점
- DB vendor에 의존하게 된다. (my sql,  h2에 의존적)
- 테이블 구조로 표현하기 때문에 엔티티가 변화했을 때 대응하기가 어렵다.

피드백을 통해서 JPA에서 쿼리를 사용할 때 어떤 방식이 있는지 학습해보게 되었습니다.

### JPQL
native query와 비슷하지만 테이블 기반이 아닌 엔티티 기반으로 쿼리가 실행된다.
Query 어노테이션을 통해 사용할 수 있고, JPQL과 native Query중 선택할 수 있다.

장점
- DB에 종속적이지 않다 (JPA에서 사용하는 쿼리문이기 때문에)
- 엔티티 기반으로 짜여지기 때문에 엔티티 변경에 효과적이다.

단점
- JOIN문 사용이 복잡하다.
- 기존의 sql과 유사하지만 문법이 달라, 사용하는데에 어색하다.
- 문자열 기반이라서 컴파일 시점에 오류를 잡기 어렵다(native query 도 동일)
- 재사용성이 어렵다.

#### Query DSL
애플리케이션 코드로 쿼리 작성이 가능하다.

장점
- 컴파일 단계에서 오류를 찾을 수 있다.
- 조건을 명시하는 등 동적으로 쿼리를 생성할 수 있다.
- 코드로 표현하기 때문에 재사용이 가능하다.

단점
- 설정이 복잡하다.
- 배우기 어렵다.

이외에도 여러가지 방식이 있는것을 알게 되었습니다. 현재의 요구사항에서는 지금처럼 직접 쿼리문을 작성해주는 것도 문제는 없어보입니다. 하지만 추후에 요구사항이 복잡해지고 엔티티 구조가 바뀐다면 지금과 같은 native query 방식은 걸림돌이 될 것 같다는 생각입니다. 그땐  러닝 커브가 존재하지만 코드로 쿼리 표현이 가능한 Query DSL 방식이 단점보단 장점이 더 많다고 느껴집니다!

아직까지 JPA에서 직접 쿼리를 사용하는 경험이 많지 않다보니, 어떤 상황에서 어떤 방식을 선택하는 것이 좋을지 확신이 서지 않는 것 같습니다. 그렇다면 샐리는 이 상황에서 JPQL이나 Query DSL과 같은 방식으로 해결하는 것이 좋다고 생각하시나요?
피드백에서 의도하신 답변이 맞을까요? 샐리의 생각이 궁금합니다!

dusdn1702 replied:
맞아요. 단순히 메서드로 제공받을 수 없는 쿼리이다보니 native query를 적용했을 것 같아서 다른 방법에 대해 여쭤봤어요.
더군다나 이렇게 `JpaRepository`를 구현하는 Repository에서 native query를 사용하게 되면 또 어떤 문제가 있을 수 있을까요?
가능하다면 JPQL도 구현해보면 좋겠어요. (QueryDSL에 비해 적용이 어렵진 않을 것이라서요)

dye0p replied:
제가 느낀 단점 이외에 `native query`를 사용했을 때의 발생할 수 있는 단점을 정리해보았습니다!

- 엔티티로 짜여진 쿼리가 아니기 때문에 반환값이 **반환 타입이 엔티티 객체가 아니라면 영속성 컨텍스트에 등록되지 않습니다**. 영속성 컨텍스트에서 관리되지 않는다면 영속성 컨텍스트에서 제공하는 기능(예: 더티체킹)을 사용할 수 없게 됩니다.
- 위에서 얘기한 것 처럼 특정 DB에 대한 의존성이 증가합니다. 추후에 DB를 변경해야하는 등의 확장성을 고려하기 어렵습니다.
- 실제 쿼리문으로 짜여져있기 때문에 복잡해보인다.

이러한 단점들을 느끼고, `JPQL`을 도입해보았습니다! 그리고 제가 느꼈던 장단점입니다.

[refactor: 인기테마를 가져오는 쿼리문 JPQL로 변경](https://github.com/dye0p/spring-roomescape-waiting/commit/cbcef6926cdeb887386c3e3ed818a274de9ebc2f)

장점
`native query` 보다 가독성이 좋다.
- JPQL은 테이블이 아닌, 엔티티를 기반으로 쿼리가 짜여지기 때문에 엔티티 필드를 사용할 수 있습니다.
- 위의 장점으로 인해서 특정 DB에 종속되지 않기 때문에 추후 확장성이 가능한 구조가 됩니다. 즉, JPA 에서 제공하는 추상화의 이점을 느낄 수 있었습니다.

단점
- `native qurey`에 비해서 비교적 다양한 기능을 제공하지 않는다.
- 위의 이유와 연결되어서 복잡한 쿼리 작성시 오히려 `native query`보다 복잡해질 수 있다.

이 부분은 단점인지 잘 모르겠습니다. 특정 DB벤더 특화 기능을 사용할 수 없습니다.  인기테마를 10개로 필터링하는 구문에서 기존에는 `LIMIT`을 사용하였지만, `JPQL`은 `LIMIT`을 지원하지 않아, `Pagble`로 대체 하였습니다! 

`JPQL`을 사용했을 때 확실히 기존의 `native query` 보다 간결해지고 테이블 대신, 엔티티를 사용하여 객체지향적으로 쿼리를 작성할 수 있음을 느꼈습니다. 직접 쿼리를 작성해야 할 때, 간단한 조회나 복잡하지 않은 쿼리의 경우 `JPQL`을 사용하는 것이 좋다고 생각합니다. 반대로 복잡한 쿼리를 작성해야 하는 경우 오히려 `native query`가 더 나을수도 있을 것 같습니다!

dusdn1702 replied:
💯 좋습니다 ㅎㅎ

dusdn1702 commented on src/main/java/roomescape/service/ReservationService.java:
```diff
@@ -0,0 +1,112 @@
+                .toList();
+    }
+
+    public void cancelReservation(Long id) {
```
삭제와 저장의 요청이 동시에 오게 되면 어떻게 처리가 될까요?

dye0p replied:
생각하지 못한 부분입니다..!

동시에 두개의 요청이 호출되면 저장과 삭제 메서드는 별도의 트랜잭션으로 동작하기 때문에 서로의 작업 상태를 알 수 없습니다. A가 예약을 삭제하고, 동시에 B가 예약을 추가하면, 둘의 순서에 따라서 결과가 달라지게 됩니다. 따라서 트랜잭션의 기본 격리 수준에서는 격리성을 충분히 보장하지 못할 수도 있습니다. 이를 해결하기 위해서 Lock 전략을 사용할 수 있는걸로 알고 있습니다.

Lock을 거는 방식에는 대표적으로 두 종류가 있습니다. 

### 낙관적 락
데이터를 저장할 때 version을 증가시킨다.
우선은, 동시충돌이 없다고 가정하고, 데이터를 조회할때 version정보도 함께 조회하여 충돌 여부를 확인한다.
`LockModeType.OPTIMISTIC` +  `@Version` 필드를 사용

장점
- 매번 동시성 처리를 하지 않기 때문에 성능에 유리할 수 있다.

단점
- 충돌 시 검사를 하기 때문에 예외가 발생한다. 
- 이미 실행 후 예외가 일어나기 때문에 재시도를 해야한다.

### 비관적 락
충돌이 일어날 거라 가정하고, 데이터를 변경하는 동안 락을 건다.
락이 걸린 트랜잭션외에 다른 트랜잭션은 읽기, 쓰기가 불가능하다
`ockModeType.PESSIMISTIC_WRITE`

장점
- 동시성 충돌이 발생하기 전에 락을 걸기 때문에 데이터 정합성을 보장할 수 있다.

단점
- 락을 거는 동안 대기시간이 발생한다.
- 데드락 발생 가능성이 있다.

만약 현재와 같은 상황(예약을 삭제하는 동시에 누군가 같은 예약을 하는 경우)이라면 동시성 충돌이 자주 일어날 수 있는 상황이라고 생각합니다. 따라서 예약을 저장하는 로직에서 예약의 존재를 확인하는 메서드에 락을 걸어줌으로 비관적 락을 사용하는 것이 낙관적 락을 사용하는 것보다 좋은 선택이라고 생각이 듭니다. 

예약 조회 메서드에 락을 걸어주는 이유로는, 사용자가 예약을 저장, 삭제하는 과정은 **예약 조회** 후에 일어납니다.
- 예약 저장 시점
   - 먼저 존재하는 예약인지 확인하는 조회
- 예약 삭제 시점
   - 삭제하려는 예약지 존재하는지 확인하는 조회   
  
위와같은 이유로 조회 시점에 락을 걸면 조회 시점에서 동시에 저장과 삭제에 접근을 못할 것이라고 생각했습니다.

비관적 락 적용 예시
``` java
@Lock(LockModeType.PESSIMISTIC_WRITE) 
Optional<Reservation> findByDateAndReservationTime(LocalDate date, ReservationTime time);
```

하지만, 이 부분에 대해서는 제대로 알지 못한 부분이라, 제대로 판단하기가 어려운것 같습니다. 사실 **충돌이 자주 일어날 수 있는 상황을 판단** 하는 것도 개인적인 부분이라 판단의 근거가 뚜렷하지 않은 것 같습니다. 이 부분에 대한 샐리의 생각은 어떤가요?

추가로 동시성 테스트를 어떻게 수행할 수 있을지 수행 방식에 대한 고민을 하였습니다. 
`@RepeatedTest`, `ExecutorService`를 통해 테스트를 구현할 수 있다는 것을 알게 되었는데, 공부를 더 해봐야 할 것 같습니다.
그 과정에서 지금까지의 테스트는 순차적 테스트를 작성했다는 것도 알게 되었네요..! 

샐리가 의도하신 답변이 맞을까요?! 피드백에서 의도한 답변이 아니거나, 잘못된 답변을 한 부분이 있다면 피드백 해주시면 감사하겠습니다 : )

dusdn1702 replied:
락까지 학습하실줄 몰랐어요. 너무 좋네요.ㅎㅎ
서버 개발자로서 동시성 이슈는 앞으로 다양한 방식으로 마주하게 되고 고민해야하는 부분이니 이번 기회에 학습해보면 좋겠어요 😄 
이번에 `@Transactional` 코드를 추가해주셨는데요, 말씀 주신 테스트 구현 방식을 통해 어노테이션을 달기 전, 추가한 후, 락 추가 이렇게 3단계로 나누어 실험해보면 어떨까요?
그 과정에서 락은 가볍게만 사용해봐도 좋을 것 같아요. 어디에 어떻게 적용할지 어렵다면 락과 관련한 부분은 추후에 학습해봐도 좋아요.

dye0p replied:
동시성과 락 관련된 부분은 학습할 거리가 많다고 느껴졌습니다! 좋은 피드백 감사합니다: ) 👍

dusdn1702 commented on src/test/java/roomescape/repository/ThemeRepositoryTest.java:
```diff
@@ -0,0 +1,87 @@
+import roomescape.model.Theme;
+
+@DataJpaTest
+class ThemeRepositoryTest {
```
repositoryTest는 왜 Theme만 있나요??

dye0p replied:
이전 미션에선 JDBC 환경에서 직접 SQL 쿼리문을 작성했기 때문에 **작성한 쿼리가 정상 동작하는지** 테스트로 검증해야 한다고 생각했습니다. 하지만 JPA로 바뀌면서 직접 작성하던 쿼리문을 JPA에게 맡기게 되었고, 결론적으로 "**sql은 JPA가 생성하고, 실행하는 것이므로 우리가 검증할 필요 없이 JPA를 신뢰 하는것이 좋겠다.**" 라는 판단을 하였습니다. 

`ThemeRepository`를 테스트한 이유는 네이티브 쿼리를 사용한 사용자 정의 메서드를 검증하기 위함이었습니다. 왜냐하면 개발자가 제대로 쿼리를 작성하였는지 검증 할 필요가 있다고 생각했기 때문입니다. 반대로 말해서 다른 `Repository`를 테스트하지 않은 이유는 JPA가 생성한 쿼리를 신뢰하기 때문 이었습니다. 

하지만 피드백을 받고 고민을 해 봤을 때 개발자가 JPA를 통해서 직접 정의해준 메서드는 비즈니스 로직을 위해서 생겨난 메서드니까 테스트 해야 할까? 라는 의문이 들었고, 테스트를 작성하는 것이 좋겠다고 생각하여 테스트를 작성했습니다!

여기서 드는 궁금증은 
- 만약 이외의 정의된 `Repository` 메서드를 테스트 해야 한다면 기본적으로 정의 되어있는 `save()`, `findById()` 같은 메서드도 테스트 해야하는지? 테스트 해야하는 상황이 생길수도 있는지?
- JPA의 메서드를 어디까지 신뢰하는 것이 좋을지?

의문입니다. 
이에 대한 샐리의 생각이 궁금합니다! 

[test: ThemeRepository 테스트 추가](https://github.com/woowacourse/spring-roomescape-waiting/pull/234/commits/86fa8de5c6beaca0d9572d29f3d47d867722ba5c)
[test: MemberRepository 테스트 추가](https://github.com/woowacourse/spring-roomescape-waiting/pull/234/commits/4ef59ddb25c2d70858d677a99a902fb97a2bb78c)
[test: ReservationRepository 테스트 추가](https://github.com/woowacourse/spring-roomescape-waiting/pull/234/commits/7c424a888151434d1566efb9da7d1f69558c3148)
[test: ReservationTimeRepository 테스트 추가](https://github.com/woowacourse/spring-roomescape-waiting/pull/234/commits/fdd16a74a32ad243019cde59170b66a959faa2da)

dusdn1702 replied:
저라면 지금 처음 jpa를 도입하고 학습하는 과정에서 repository도 사용한 메서드들은 검증해볼 것 같아요.
그 과정에서 jpa 메서드에 대한 신뢰가 생기는 것이니까요. 내가 구현한 것이 아닌데 신뢰는 어디서 나오나요?
추후에 운영하는 코드에서 버전이 변경되면서 save 내부 로직이 변경됐을 때 테스트를 통해 단번에 코드의 변경을 인지할 수 있으면 좋지 않을까요?

dye0p replied:
피드백 감사합니다 : ) 학습의 측면에서 repository를 테스트 하는 것은 동의합니다! 👍

코멘트에서 몇가지 이해가 잘 되지 않는 부분이 있어요!
> 내가 구현한 것이 아닌데 신뢰는 어디서 나오나요?

직접 구현하지 않았더라도, `findById()`, `save()`와 같은 로직은 자바에서 제공하는 `toString()`, `equals() & hashCode()` 처럼 기본적으로 제공하는 메서드라고 인식하고 있습니다, 저는 자바에서 기본적으로 제공하는 라이브러리나 메서는 신뢰를 하고 사용하는 편입니다! 그렇다면 **jpaRepository에서 제공하는 기본적인 메서드는 신뢰해도 되지 않을까?** 라는 생각입니다!  만약 기본적으로 제공하는 메서드를 신뢰하기 어렵다면 **자바에서 제공하는 모든 메서드도 테스트 해야할까?** 라는 의문이 드는 것 같습니다!

제가 궁금한 부분에 대해서 이 부분을 고려하고 말씀하신게 맞을까요?! 제가 이해를 잘 못한 부분이 있다면 지적해주세요! 🙇‍♂️

> 추후에 운영하는 코드에서 버전이 변경되면서 save 내부 로직이 변경됐을 때 테스트를 통해 단번에 코드의 변경을 인지할 수 있으면 좋지 않을까요?

이 코멘트에서 말씀하시는 save 로직은 japRepository 에서 기본 내장되어있는 테스트를 말씀하신게 맞을까요?? 
물론 save 와 같은 로직의 내부가 변경 되었을 때 테스트를 통해서 인지할 수 있다면 좋을 것 같습니다! 하지만 이것도 위와같은 이유와 같은 맥락으로 아직까지 잘 모르겠습니다..ㅜ 

샐리의 피드백에 저의 의견을 남겨봤습니다! 제가 피드백을 잘 못 이해하고 있을 수도 있다는 생각도 드네요!

dusdn1702 replied:
이해하신 부분이 맞아요!
저는 java는 언어인 반면에 jpa의 경우 구현체를 도입했다고 생각해요. `toString(), equals() & hashCode()` 처럼 단순한 로직도 아니죠.
이 과정에서 로직이 변경될 여지가 있기 때문에 테스트로 검증할 수 있다면 테스트하는 것이 좋다고 생각해요.

dusdn1702 reviewed (changes requested):
안녕하세요 도기~
jpa 미션 구현 잘해주셨네요. 고생 많으셨습니다!
남겨둔 코멘트 확인 부탁드릴게요 ㅎㅎ

dye0p committed:
refactor: 엔티티 컬럼 매핑 제약조건 설정

- 자동 생성되는 테이블의 제약조건을 설정하기 위해서 @Column, @JoinColumn 을 사용하여 설정

dye0p committed:
refactor: 연관관계 로딩 전략 번경(즉시 로딩 -> 지연 로딩)

- 필요하지 않은 시점에 테이블이 조회되는 것을 방지하기 위해서 ManyToOne 연관관계를 맺는 필드의 로딩 전략을 지연 로딩으로 변경하였음.

- 지연로딩으로 인해 필요한 시점에 쿼리를 발생시키기 때문에 제대로된 동작을 하지 않을수가 있어, repository를 호출하는 서비스 로직에 트랜잭션 어노테이션을 사용하여 트랜잭션 처리를 해주었음

dye0p committed:
refactor: 사용하지 않는 AllArgsConstructor 어노테이션 제거

dye0p committed:
refactor: repository 메서드 네이밍 컨벤션에 맞게 수정

dye0p committed:
test: ThemeRepository 테스트 추가

dye0p committed:
test: MemberRepository 테스트 추가

dye0p committed:
test: ReservationRepository 테스트 추가

dye0p committed:
test: ReservationTimeRepository 테스트 추가

dye0p committed:
fix: 이미 예약 되어있는 테마를 삭제할 때 커스텀 예외가 발생하지 않던 문제 수정

- 기존의 로직에서 트랜잭션의 커밋 범위가 catch문을 벗어나서 일어나기 때문에 커스텀 예외를 발생시킬 수 없고, 서버에러로 넘어가버린다. 이를 해결하기 위해서 테마를 삭제하기 전, 예약 테이블에 테마가 존재하는지 먼저 검증 하는 로직을 추가하였다.

dye0p reviewed (commented):
안녕하세요 샐리! JPA의 지식이 미흡해서 코멘트에 대해서 코드 변경보단, 공부 하는 시간이 길었던 것 같습니다. 그래서 피드백 반영이 조금 늦어졌습니다🥹 리뷰 잘 부탁 드립니다 : )

dye0p commented:
@CodeRabbit 작성한 코드에 대해서 전체적으로 리뷰를 작성해줘, JPA를 통해서 코드가 객체지향적으로 짜여졌는지, 테스트 방식은 어떤지 평가해줘

dusdn1702 commented on src/main/java/roomescape/model/ReservationTime.java:
```diff
@@ -0,0 +1,28 @@
+    private Long id;
+
+    @Column(nullable = false)
+    private LocalTime startAt;
```
LocalTime을 jpa는 어떻게 변환해서 저장할까요??

dye0p replied:
이 부분에 대해서 큰 고민을 안해본 것 같네요! JPA에서 date나 time 같은 타입을 자동으로 변환하여 db에 매핑시키는걸로 생각했습니다! 잘 모르고 있었다 생각했고, 조금 더 알아봤습니다!

우선 Hibernate User Guide에 따르면, java.time.LocalTime은 JDBC TIME 타입으로 기본 처리되며, LocalDateTime은 TIMESTAMP 등으로 매핑된다고 합니다!

기본적인 동작 과정은 아래와 같이 정리할 수 있습니다!
- JPA는 엔티티의 Java 타입 필드를 
JDBC를 통해 SQL 데이터 타입으로 변환하여 DB에 저장한다.

이 변환은 JPA가 내부적으로 JDBC 타입 매핑 또는 `AttributeConverter`, 기본 Java 타입 매핑을 사용해서 처리합니다.

```java
@Column(nullable = false)
private LocalTime startAt;
```
Hibernate는 LocalTime을  JDBC의 java.sql.Time 또는 java.time.LocalTime과 연동하고, sql의 TIME 타입으로 변환해서 저장합니다. 

이게 가능한 이유는 Hibernate5.2 버전부터 Hibernate-core에 추가되어서 별도의 어노테이션 없이 매핑이 가능했습니다.

Hibernate5.2 이하 버전에서는  자바의 time 타입을 기본적으로 매핑할 수 없었습니다. 그래서 지금처럼 엔티티의 필드에 LocalTime 타입으로 선언하면 예외가 발생했습니다. 이를 매핑시켜주기 위해서 `@AttributeConverter` 등과 같은 어노테이션을 사용할 수 있었습니다.

결론적으로 엔티티에 명시된 필드 타입에 따라서 JPA 내부에서 자동으로 매핑을 하여 db에 저장됩니다!

dusdn1702 commented on src/main/java/roomescape/presentation/resvoler/LoginMemberArgumentResolver.java:
```diff
@@ -0,0 +1,42 @@
+package roomescape.presentation.resvoler;
```
오타가 있네요. ㅎㅎ

dye0p replied:
수정했습니다! 👍

[style: 패키지명 오타 수정(resvoler -> resolver)](https://github.com/dye0p/spring-roomescape-waiting/commit/bf93159667a5644053758821994d196d2a3f5008)

dusdn1702 commented on src/main/java/roomescape/repository/ReservationTimeRepository.java:
```diff
@@ -0,0 +1,12 @@
+import roomescape.model.ReservationTime;
+
+@Repository
+public interface ReservationTimeRepository extends JpaRepository<ReservationTime, Long> {
```
Repository와 Model을 분리해주셨는데요, model이 모두 `@Entity`로 선언되어 둘은 매우 밀접해보여서요. 패키지를 나눈 이유가 무엇인가요?

dye0p replied:
처음 패키지를 나눴던 기준은 단순한 이유로, domain과 repository는 별개의 계층이라고 생각했기 때문입니다. repository는 엔티티를 db에 저장하는 역할을 하기 때문이라고 생각했습니다. 

사실
> model이 모두 `@Entity`로 선언되어 둘은 매우 밀접해보여서요

이 코멘트를 받기 전까지 이것에 대해서 깊게 생각 해본적 없었습니다! 
이전, JDBC를 사용할 땐 개발자가 직접 db에 값을 삽입시키는 로직을 구현했습니다. 그래서 저희가 작성한 코드는 값을 db에 저장하는 과정을 코드로 추상화 했다고 생각합니다. 
그래서 dao 객체는 영속성 계층에 속하는 것이 맞다고 판단했습니다. 

JPA를 사용하면서 **repsotiry**도 같은 개념이라고 생각했습니다. 단순하게 "**기존의 직접짜던 쿼리를 JPA가 수행해주는 구나**" 라고 생각했습니다. 그래서 피드백을 받고, 여러가지 자료를 찾아 봤을 때 repository를 바라보는 관점이 여러가지가 있다는 것을 알게 되었습니다.

DDD 관점에서 repository를 일급 컬렉션으로 바라보는 repository pattern을 봤을 때, 맞는 말 인것 같았지만, 뭔가 와닿지가 않았습니다. 하지만 말씀해주신 것 처럼 **repository**는 현재 엔티티와 밀접하다고 생각합니다. 각각 엔티티의 타입에 맞는 **repository**를 가지고 있으니까요!

하지만 JPA **repository**를 영속성의 도구로 본다면 계층을 분리하는 것도 납득이 될 것 같습니다.

**repository**와 **domain** 모델을 어떻게 바라보는 것이 좋을까요? 샐리는 어떻게 생각하시나요?

dusdn1702 replied:
지금은 도메인 모델 분리가 불필요해서 이러한 상황이 나왔다고 생각해요.
저는 repoistory 와 entity 는 jpa에 연관된 같은 계층,
domain은 비즈니스 로직이 들어가는 다른 계층으로 바라보는 편이에요.

dusdn1702 reviewed (changes requested):
안녕하세요 도기~
길지 않은 시간임에도 피드백 하나하나에 깊이 있게 학습해주신 점이 매우 인상 깊었습니다. 고생 많으셨어요.
코드에 대한 수정 보다는 추가적으로 생각해봄직한 코멘트 추가로 남겨두었어요. 확인 부탁드릴게요 !

dye0p committed:
refactor: getter가 필요한 필드에만 getter 적용

dye0p committed:
style: 패키지명 오타 수정(resvoler -> resolver)

dye0p committed:
refactor: 인기테마를 가져오는 쿼리문 JPQL로 변경

dye0p reviewed (commented):
안녕하세요 샐리! 피드백 반영하여 리뷰 요청드립니다! 잘 부탁드립니다! : )

dusdn1702 reviewed (approved):
안녕하세요 도기
이번 피드백도 깊이 있는 학습 해주셔서 감사합니다. 고생 많으셨어요.
미션에서 학습할 내용들은 충분히 수행해주신 것 같아 이번 미션은 여기서 마무리하도록 하겠습니다.
다음 미션에서 뵐게요!
